File: php/src/CodingInterview/Application/Usecase/Order/RebalanceOrderUsecase.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Usecase\Order;

use Folio\CodingInterview\Application\Repository\AccountRepository;
use Folio\CodingInterview\Application\Repository\PortfolioRepository;
use Folio\CodingInterview\Domain\UserId;

final class RebalanceOrderUsecaseInput
{
    public function __construct(public readonly UserId $userId) {}
}

final class RebalanceOrderUserNotFoundException extends \RuntimeException
{
    public function __construct() { parent::__construct('user has no live account'); }
}

final class RebalanceOrderUsecase
{
    public function __construct(
        private readonly AccountRepository $accountRepository,
        private readonly PortfolioRepository $portfolioRepository,
    ) {}

    public function run(RebalanceOrderUsecaseInput $input): void
    {
        $account = $this->accountRepository->find($input->userId);
        if ($account === null) {
            throw new RebalanceOrderUserNotFoundException();
        }
        $portfolio = $this->portfolioRepository->get();
        $updated = $account->rebalance($portfolio);
        $this->accountRepository->upsert($input->userId, $updated);
    }
}

File: golang/internal/application/usecase/order/rebalance_order_usecase.go
package order

import (
	"errors"

	"folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/domain"
)

var ErrRebalanceUserNotFound = errors.New("user has no live account")

type RebalanceOrderUsecaseInput struct {
	UserId domain.UserId
}

type RebalanceOrderUsecase struct {
	accountRepo   repository.AccountRepository
	portfolioRepo repository.PortfolioRepository
}

func NewRebalanceOrderUsecase(
	accountRepo repository.AccountRepository,
	portfolioRepo repository.PortfolioRepository,
) *RebalanceOrderUsecase {
	return &RebalanceOrderUsecase{
		accountRepo:   accountRepo,
		portfolioRepo: portfolioRepo,
	}
}

func (u *RebalanceOrderUsecase) Run(input RebalanceOrderUsecaseInput) error {
	account, err := u.accountRepo.Find(input.UserId)
	if err != nil {
		return err
	}
	if account == nil {
		return ErrRebalanceUserNotFound
	}

	portfolio, err := u.portfolioRepo.Get()
	if err != nil {
		return err
	}

	updated := account.Rebalance(portfolio)

	return u.accountRepo.Upsert(input.UserId, updated)
}
